backend/x86_peephole/x86_peephole_rules.ml

```diff
@@ -0,0 +1,437 @@
+        result
+      | _, _, _, _ -> U.No_match)
+    | Ins (MOV (src1, dst1)), Ins (SUB (src2, dst2)) when U.equal_args dst1 dst2
+      -> (
+      (* Pattern: mov %a, %b; sub $CONST, %b *)
+      match src1, dst1, src2, dst2 with
+      | Reg64 a, Reg64 b, Imm imm, Reg64 b'
+        when U.equal_reg64 b b'
+             (* For SUB, we negate CONST, so we need CONST != Int32.min_int to
+                avoid overflow *)
+             && Int64.compare imm (Int64.of_int32 Int32.min_int) > 0
+             && Int64.compare imm (Int64.of_int32 Int32.max_int) <= 0 ->
+        (* Negate the immediate: sub $CONST, %b becomes lea -CONST(%a), %b *)
+        let displ = Int64.to_int (Int64.neg imm) in
+        let result = apply_lea_optimization cell1 cell2 dst1 a None displ in
+        (match result with
+        | U.Matched _ ->
+          stats.rewrite_mov_add_to_lea <- stats.rewrite_mov_add_to_lea + 1
+        | U.No_match -> ());
+        result
+      | _, _, _, _ -> U.No_match)
+    | Ins (MOV (src1, dst1)), Ins (INC dst2) when U.equal_args dst1 dst2 -> (
+      (* Pattern: mov %a, %b; inc %b *)
+      match src1, dst1, dst2 with
+      | Reg64 a, Reg64 b, Reg64 b' when U.equal_reg64 b b' ->
+        let result = apply_lea_optimization cell1 cell2 dst1 a None 1 in
+        (match result with
+        | U.Matched _ ->
+          stats.rewrite_mov_add_to_lea <- stats.rewrite_mov_add_to_lea + 1
+        | U.No_match -> ());
+        result
+      | _, _, _ -> U.No_match)
+    | Ins (MOV (src1, dst1)), Ins (DEC dst2) when U.equal_args dst1 dst2 -> (
+      (* Pattern: mov %a, %b; dec %b *)
+      match src1, dst1, dst2 with
+      | Reg64 a, Reg64 b, Reg64 b' when U.equal_reg64 b b' ->
+        let result = apply_lea_optimization cell1 cell2 dst1 a None (-1) in
+        (match result with
+        | U.Matched _ ->
+          stats.rewrite_mov_add_to_lea <- stats.rewrite_mov_add_to_lea + 1
+        | U.No_match -> ());
+        result
+      | _, _, _ -> U.No_match)
+    | _, _ -> U.No_match)
+  | _ -> U.No_match
+
+(* Rewrite rule: optimize MOV followed by ADD (register) to LEA. Pattern: mov
+   %a, %b; add %c, %b Rewrite: lea (%a,%c), %b
+
+   This is safe when: - All operands are Reg64 registers - %b ≠ %c (if %b == %c,
+   ADD would use the new value of %b after MOV, but LEA would use the old value)
+   - Flags written by ADD are dead (LEA doesn't modify flags)
+
+   Note: SUB cannot be optimized here because LEA computes base + index * scale
+   + disp where scale ∈ {1,2,4,8}. To compute %a - %c would require scale = -1,
+   which is not supported by x86-64.
+
+   Key difference: ADD sets flags (CF, OF, SF, ZF, AF, PF), LEA doesn't. *)
+let rewrite_mov_add_reg_to_lea stats cell =
+  (* CR xclerc: check this rewrite is always beneficial. *)
+  match U.get_cells cell 2 with
+  | [cell1; cell2] -> (
+    match DLL.value cell1, DLL.value cell2 with
+    | Ins (MOV (src1, dst1)), Ins (ADD (src2, dst2)) when U.equal_args dst1 dst2
```

Comment:
`when U.equal_args` and `when U.equal_reg64` are redundant, right? (also applies to previous function)